Fix/7/compiled binary in archive

curzolapierre opened the pull request:
Fix #7
Documentation: https://www.notion.so/scalingooriginal/New-GraphicsMagick-release-dc897365130545a8892a942b9a30eff1

curzolapierre committed:
fix(package): update package_graphicsmagick to build the binary

curzolapierre committed:
fix(compile): update compile to work with the binary now

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,103 @@
+# if [ "$S3_BUCKET" == "" ]; then
+#   echo "must set S3_BUCKET"
+#   exit 1
+# fi
```
```suggestion
```

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,103 @@
 
 if [ "$STACK" == "" ]; then
-  echo "must set STACK to the heroku stack being compiled for (i.e. STACK=cedar or STACK=cedar-14)"
+  echo "must set STACK to the scalingo stack being compiled for (i.e. STACK=scalingo-18)"
```
nitpick: let's suggest the most recent stack ;)
```suggestion
  echo "must set STACK to the scalingo stack being compiled for (i.e. STACK=scalingo-20)"
```

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,103 @@
-echo "    from ${gm_local_path}"
+# echo ">>> Uploading graphics magick"
+# echo "    to ${S3_BUCKET}/${REMOTE_PATH}"
+# echo "    from $gm_local_path"
```
```suggestion
```

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,103 @@
-$basedir/aws/s3 put $S3_BUCKET $STACK/$REMOTE_PATH $gm_local_path
+tar -czf $gm_local_path -C $vendored_graphicsmagick_dir .
+echo -e "You can upload archive: $gm_local_path to S3 now: \n\t$gm_local_path"
+# ${basedir}/aws/s3 put $S3_BUCKET ${STACK}/${REMOTE_PATH} $gm_local_path
```
```suggestion
```

EtienneM reviewed (changes requested):
Don't you also want to make use of the cache for the archive? For each build there is a cache between deployments. We talk about it here: https://doc.scalingo.com/platform/deployment/buildpacks/custom#architecture-of-a-buildpack

If you use the cache for the GraphicsMagick archive  in `bin/compile`, you don't need to download the archive anymore after the first deployment!

curzolapierre committed:
Apply suggestions from code review

Co-authored-by: Étienne M. <EtienneM@users.noreply.github.com>

curzolapierre committed:
fix(compile): add cache management

EtienneM commented on bin/compile:
```diff
@@ -42,25 +44,37 @@ echo "Installing graphicsmagick $GRAPHICS_MAGICK_VERSION" | arrow
+  # vendor directories
+  mktmpdir $graphicsmagick_cache_dir
+  echo "Downloading archive in $graphicsmagick_cache_dir" | arrow
+  archive_download "$graphicsmagick_cache_dir"
```
I think this should be split in two different conditions:
1. First you test if the cache directory exists and create it if needed
2. And then, you test if the archive is already in the cache directory, and download it if needed. 

Do you see what I mean?

EtienneM commented on bin/compile:
```diff
@@ -42,25 +44,37 @@ echo "Installing graphicsmagick $GRAPHICS_MAGICK_VERSION" | arrow
+cp ${graphicsmagick_cache_dir}/utilities/gm "${BUILD_DIR}/bin/"
 
-cp -R $VENDORED_GRAPHICSMAGICK/* $gm_lib_path
+rm -rf $VENDORED_GRAPHICSMAGICK/doc
```
`VENDORED_GRAPHICSMAGICK` is not set anymore. Do you mean `graphicsmagick_cache_dir`? In any case, I think I would prefer you test the existence of the folder `doc` before trying to delete it instead of using the `-f` flag

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,84 @@
-# build and package imagemagick for heroku
-cd $tempdir/GraphicsMagick-${GRAPHICS_MAGICK_VERSION}
-./configure --prefix /app/vendor/graphicsmagick && make install
+# we need this directory, it's our destination
```
I don't understand this comment

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,84 @@
-cd /app/vendor/graphicsmagick && tar -czvf $gm_local_path .
-$basedir/aws/s3 put $S3_BUCKET $STACK/$REMOTE_PATH $gm_local_path
+tar -czf $gm_local_path -C $vendored_graphicsmagick_dir .
+echo -e "You can upload archive: $gm_local_path to S3 now: \n\t$gm_local_path"
```
Why don't your script automatically upload the archive?

curzolapierre committed:
fix(compile): clearer way to manage the cache

curzolapierre committed:
fix(package): add a script to upload to a s3

EtienneM commented on bin/compile:
```diff
@@ -42,25 +44,43 @@ echo "Installing graphicsmagick $GRAPHICS_MAGICK_VERSION" | arrow
-make
-make DESTDIR=$gm_lib_path install
-popd
+# Test if cache directory exists
+if [ ! -d "$graphicsmagick_cache_dir" ]; then
+  # create vendor directories
+  mktmpdir $graphicsmagick_cache_dir
+fi
+
+# Test if the archive is present in the cache directory then download it if not.
+if [ ! -d "${graphicsmagick_cache_dir}/utilities" ]; then
+  echo "Downloading archive in $graphicsmagick_cache_dir" | arrow
+  archive_download "$graphicsmagick_cache_dir"
+fi
```
Sorry I missed it during the previous reviews... But the cache should be invalidated if you update the GraphicsMagick version (`GRAPHICS_MAGICK_VERSION`). By reading the code I feel like it does not get invalidated if the version changes. We usually put the version in the cache folder name so that the cache gets invalidated when the version changes. Do you see what I mean?

EtienneM commented on bin/compile:
```diff
@@ -42,25 +44,43 @@ echo "Installing graphicsmagick $GRAPHICS_MAGICK_VERSION" | arrow
+
+# If doc folder still exitst, we can delete it safely
+if [ -d "${graphicsmagick_cache_dir}/doc" ]; then
+  rm -rf ${graphicsmagick_tmp_dir}/doc
```
The `f` option shouldn't be mandatory, I don't like having it
```suggestion
  rm -r ${graphicsmagick_tmp_dir}/doc
```

EtienneM reviewed (changes requested):
Just a few remaining comments, but thanks for the update! Did you already write the Notion page? I would like to reveiw it alongside this PR

curzolapierre commented:
The notion page is here : https://www.notion.so/scalingooriginal/New-GraphicsMagick-release-dc897365130545a8892a942b9a30eff1

curzolapierre committed:
fix(cache): invalid the cache in version change

EtienneM commented on bin/compile:
```diff
@@ -42,25 +44,43 @@ echo "Installing graphicsmagick $GRAPHICS_MAGICK_VERSION" | arrow
+
+# If doc folder still exitst, we can delete it safely
+if [ -d "${graphicsmagick_cache_dir}/doc" ]; then
+  rm -fr ${graphicsmagick_tmp_dir}/doc
```
```suggestion
  rm -r ${graphicsmagick_tmp_dir}/doc
```

EtienneM commented on support/package_graphicsmagick:
```diff
@@ -1,120 +1,93 @@
+make DESTDIR=$gm_lib_path install
+popd
+
+echo ">>> Creating graphics magick archive"
```
nitpicky: this is really me being picky but I don't like the fact that GraphicsMagick is spelled with various different spellings in the same file that bother me. I tihnk it should be `GraphicsMagick` everywhere

EtienneM reviewed (changes requested):
I also commented the Notion page. Can you upload to S3 for both stacks we currently support?

curzolapierre commented:
> Can you upload to S3 for both stacks we currently support?

Done

curzolapierre committed:
fix(s3): vendor are downloaded from graphics-magick sources directly

Upload to s3 is now correctly done under buildpacks-repository bucket under graphicsmagick directory.

EtienneM reviewed (approved):
Thanks a lot 👍🏾
